Feature/54733 primerise the activity panel

jjabari-op opened the pull request:
https://community.openproject.org/projects/openproject/work_packages/54733

jjabari-op committed:
[#54733] Primerise the Activity panel

https://community.openproject.org/work_packages/54733

jjabari-op committed:
introducing primerized work package activities

jjabari-op committed:
Merge branch 'dev' into feature/54733-primerise-the-activity-panel

jjabari-op committed:
WIP: primerizing work_package activities, editing not yet implemented, tagging users not yet possible

jjabari-op committed:
minor cleanup

jjabari-op committed:
activate primierized activities in pullpreview env

jjabari-op committed:
enable editing

jjabari-op committed:
prevent edit state form being overwritten and added basic polling for evaluation as discussed with @wielinde

jjabari-op committed:
Merge branch 'dev' into feature/54733-primerise-the-activity-panel

jjabari-op committed:
changed polling frequency to 1 minute

jjabari-op committed:
added comments only filter

jjabari-op committed:
added i18n

jjabari-op committed:
highlight journal items with unread notification, minor refactoring

jjabari-op committed:
changed filter component as seen on updated mockups

jjabari-op committed:
added sorting update directly in activity tab

jjabari-op committed:
implemented sticky bottom editor approach, file upload and mentions not working yet via ckeditor

jjabari-op committed:
group journals by day

jjabari-op committed:
optimized syncing approach

jjabari-op committed:
implemented dynamic heights/margins of editor and journals container

jjabari-op committed:
Merge branch 'dev' into feature/54733-primerise-the-activity-panel

jjabari-op committed:
minor adjustments towards resource injection, not restricting the editor type

jjabari-op committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/54733-primerise-the-activity-panel' into feature/54733-primerise-the-activity-panel

jjabari-op committed:
support mentions and uploads based on feedback from @oliverguenther

jjabari-op committed:
added quoting

jjabari-op committed:
Merge branch 'dev' into feature/54733-primerise-the-activity-panel

jjabari-op committed:
fixed sass imports

jjabari-op committed:
fixed bg colors of inputs

jjabari-op committed:
close editor on blur and keep editor open after submit for a more chat like interface as requested by @wielinde

jjabari-op committed:
fixed blur event listener

jjabari-op committed:
decreased timeout as much as possible

jjabari-op committed:
fixed rendering after sorting update

jjabari-op committed:
mobile adjustments following iteration made for the mobile concept

jjabari-op committed:
adjusted max height of editor as requested and fixed margin adjustment

jjabari-op committed:
decreased polling interval as requested by @wielinde, optimization based on user activtity pending

jjabari-op committed:
fixed paddings

jjabari-op committed:
optimized polling approach in order to avoid unnecessary requests

jjabari-op committed:
avoid editor data loss on tab change or other browser events unmounting the stimulus component

jjabari-op committed:
implemented github PR style

jjabari-op committed:
fixed detail update streams

jjabari-op committed:
fixed filter behavior

jjabari-op committed:
quickly adjusted font rendering of detail descriptions, refactoring required

jjabari-op committed:
Merge branch 'dev' into feature/54733-primerise-the-activity-panel

jjabari-op committed:
added first specs for primerized workpackage activities

jjabari-op committed:
implemented feedback from product team and adjusted micro behaviors, needs refactoring

jjabari-op committed:
Merge branch 'dev' into feature/54733-primerise-the-activity-panel

jjabari-op committed:
fixed bottom detection for md viewports

jjabari-op committed:
Merge branch 'dev' into feature/54733-primerise-the-activity-panel

jjabari-op committed:
implemented feedback from product team, finalized new specs, cleanup, bugfixing

jjabari-op committed:
cleanup temp test code

jjabari-op committed:
Merge branch 'dev' into feature/54733-primerise-the-activity-panel

jjabari-op committed:
fixing eslint issues

jjabari-op committed:
enable primerized workpackages feature flag in ci

jjabari-op committed:
reenabled and fixed failing spec

jjabari-op committed:
refactored to resolve rubocop issues

jjabari-op committed:
more refactorings to resolve rubocop issues

jjabari-op committed:
fixed ckeditor bottom margin

jjabari-op committed:
hide stem connection when empty

jjabari-op committed:
refactored and fixed stem behaviour for filter states

oliverguenther commented on frontend/src/app/features/work-packages/components/work-package-comment/work-package-comment.component.html:
```diff
@@ -1,4 +1,18 @@
+      <svg:rect x="0" y="10" width="100" height="8" rx="1" />
+  
+      <svg:rect x="0" y="25" width="100" height="12" rx="1" />
+    </op-content-loader>
```
The content-loader doesn't really match the appearance of the tab itself. It seems to be copied from the meeting tab? Either we provide it's own appearance, or we abstract this appearance into a preset/own component

<img width="1105" alt="Bildschirmfoto 2024-08-27 um 15 23 03" src="https://github.com/user-attachments/assets/ac1ba63e-0171-40a2-a96d-0a3f0270a955">

akabiru replied:
https://community.openproject.org/wp/58459

oliverguenther commented:
- [x] There's a large between comments when filtering for comments only, that I don't see specified in Figma.
<img width="1096" alt="Bildschirmfoto 2024-08-27 um 15 30 09" src="https://github.com/user-attachments/assets/79fc4797-c36a-4fec-bb8c-5467ab6472c8">

oliverguenther commented:
- [x] Copying and visiting a link to an activity scrolls to it correctly, but that results in the comment container being scrolled weirdly:

<img width="541" alt="Bildschirmfoto 2024-08-27 um 15 31 09" src="https://github.com/user-attachments/assets/5b7e95e1-9d8f-4e45-a98e-7b2a3010d151">

oliverguenther commented:
- [ ] Hitting the action menus on mobile result in wrong placement (the known position problem)
<img width="469" alt="Bildschirmfoto 2024-08-27 um 15 38 58" src="https://github.com/user-attachments/assets/dac03439-3859-47cd-9378-60d36f6d1ff2">

oliverguenther commented:
- [ ] There is no way to cancel writing a comment (it's not specified in the ticket as far as I can tell)

oliverguenther commented on app/components/work_packages/activities_tab/index_component.html.erb:
```diff
@@ -0,0 +1,44 @@
+<%=
+  content_tag("turbo-frame", id: "work-package-activities-tab-content") do
+    component_wrapper(data: wrapper_data_attributes) do
```
Why is there turbo-frame and component_wrapper? Couldn't you use component_wrapper(tag: "turbo-frame") ?

jjabari-op replied:
No, otherwise the turbo-STREAM update of this component would include a turbo-FRAME

oliverguenther commented on app/components/work_packages/activities_tab/index_component.html.erb:
```diff
@@ -0,0 +1,44 @@
+        activties_tab_container.with_row(flex_layout: true, mt: 3) do |journals_wrapper_container|
+          journals_wrapper_container.with_row(
+            id: "journals-container",
+            classes: "with-initial-input-compensation",
```
This class should follow our BEM structure

jjabari-op replied:
Done

oliverguenther commented on app/components/work_packages/activities_tab/index_component.html.erb:
```diff
@@ -0,0 +1,44 @@
+          end
+          journals_wrapper_container.with_row(
+            id: "input-container",
+            classes: "sort-#{journal_sorting}", # css order attribute will take care of correct position
```
This too should be a BEM modifier

jjabari-op replied:
Done

oliverguenther commented on app/components/work_packages/activities_tab/index_component.rb:
```diff
@@ -0,0 +1,74 @@
+      end
+
+      def polling_interval
+        ENV["WORK_PACKAGES_ACTIVITIES_TAB_POLLING_INTERVAL_IN_MS"] || 10000
```
Remove this ENV, if it's configurable, add a Setting instead

jjabari-op replied:
I'm using this ENV in the capybara specs in order to speed up the polling and be able to test the polling behavior much faster. I don't think it's an product requirement to have this as a setting, although @wielinde expressed his interest in playing around with some values and test how it performs in real life.

If it's just for the specs, I can go for a simple `Rails.env.test?` condition

jjabari-op replied:
Only for test env now

oliverguenther commented on app/components/work_packages/activities_tab/index_component.sass:
```diff
@@ -0,0 +1,34 @@
+            right: 0
+        @media screen and (max-width: $breakpoint-xl)
+            &.sort-desc
+                order: -1
```
This whole file needs to be reworked to match our BEM css rules (https://www.openproject.org/docs/development/style-guide/frontend/#bem-css)

jjabari-op replied:
Done

oliverguenther commented on app/components/work_packages/activities_tab/journals/index_component.sass:
```diff
@@ -0,0 +1,14 @@
+            margin-top: 20px
+            height: 100vh
+        @media screen and (max-width: $breakpoint-xl)
+            display: none
```
Same as above, this should be a BEM css module

jjabari-op replied:
Done

oliverguenther commented on app/components/work_packages/activities_tab/journals/item_component.rb:
```diff
@@ -0,0 +1,153 @@
+          @filter = filter
+        end
+
+        def content
+          case state
+          when :show
+            render(WorkPackages::ActivitiesTab::Journals::ItemComponent::Show.new(**child_component_params))
+          when :edit
+            render(WorkPackages::ActivitiesTab::Journals::ItemComponent::Edit.new(**child_component_params))
+          end
+        end
```
```suggestion
        def content_component
          case state
          when :show
            WorkPackages::ActivitiesTab::Journals::ItemComponent::Show
          when :edit
            WorkPackages::ActivitiesTab::Journals::ItemComponent::Edit
          end
        end
```

instead of rendering in here, why not simply do the class check and render it in the view instead?

jjabari-op replied:
I put it into the view

oliverguenther commented on app/components/work_packages/activities_tab/journals/item_component.rb:
```diff
@@ -0,0 +1,153 @@
+        end
+
+        def editable?
+          journal.user == User.current
```
Right now it's possible to edit other user's comment. Just something to keep in mind if we want to change this behavior explicitly.

jjabari-op replied:
I've adjusted the condition to match the existing behavior: A user with the `edit_work_package_notes` permission can edit all comments, a user with only the `edit_own_work_package_notes` permission can only edit own comments

oliverguenther commented on app/components/work_packages/activities_tab/journals/item_component.rb:
```diff
@@ -0,0 +1,153 @@
+          end
+        end
+
+        def bubble_html
+          "
+          <span
+            class=\"comments-number--bubble op-bubble op-bubble_mini\"
+            data-test-selector=\"user-activity-bubble\"
+          ></span>
+          ".html_safe # rubocop:disable Rails/OutputSafety
+        end
```
```suggestion
        def bubble_html
          content_tag(:span, "", class: "comments-number--bubble op-bubble op-bubble_mini", data: { "test-selector": "user-activity-bubble")
        end
```

Why not use a Primer::Beta::Counter for this? If you need to use an op-bubble, avoid using `.html_safe`.

jjabari-op replied:
I'm using the `dot-fill` icon now

oliverguenther commented on app/components/work_packages/activities_tab/journals/item_component.sass:
```diff
@@ -0,0 +1,7 @@
+.work-packages-activities-tab-journals-item-component
+    action-menu
+        button
+            padding-top: 3px // for whatever reason, the dots in the action menu are not perfectly aligned in center vertically by default
```
This should be generalized if needed, or fixed without tag name selectors. (BEM module)

jjabari-op replied:
Fixed without tag name selector and noted down to discuss with Henriette

oliverguenther commented on app/components/work_packages/activities_tab/journals/item_component/details.rb:
```diff
@@ -0,0 +1,208 @@
+            data-test-selector=\"user-activity-bubble\"
+          ></span>
+          ".html_safe # rubocop:disable Rails/OutputSafety
+        end
```
Same as above, also extract this into a separate component if you use it multiple times

jjabari-op replied:
Done

oliverguenther commented on app/components/work_packages/activities_tab/journals/item_component/details.sass:
```diff
@@ -0,0 +1,38 @@
+            font-style: normal
+            color: var(--fgColor-muted, var(--color-fg-subtle))
+    .bubble-container
+        padding-top: 2px
```
Same as above, BEM module

jjabari-op replied:
Done

oliverguenther commented on app/components/work_packages/activities_tab/journals/item_component/details.sass:
```diff
@@ -0,0 +1,38 @@
+        transform: translateY(-50%)
+    .journal-detail-description
+        // quick hack to adapt the current detail rendering to desired primerised design
+        i
```
Avoid overriding the styles like this, it will not be maintainable. If you need help figuring out the expected behavior, reach out to @HDinger or @bsatarnejad

jjabari-op replied:
Noted. I did this because I was told to not touch the rendering of the journal detail descriptions. So I don
t have control over the HTML rendered in these lines. I will ask Henriette how to better handle this

oliverguenther commented on app/components/work_packages/activities_tab/journals/item_component/edit.html.erb:
```diff
@@ -0,0 +1,35 @@
+                     size: :medium,
+                     tag: :a,
+                     href: cancel_edit_work_package_activity_path(work_package.id, id: journal.id, filter:),
+                     data: { "turbo-stream": true }
```
Nitpick: You're mixing `data: { turbo_stream: true } ` and `data: { "turbo-stream": true }`. I'd prefer to use turbo_stream: true

jjabari-op replied:
Done

oliverguenther commented on app/components/work_packages/activities_tab/journals/item_component/edit.sass:
```diff
@@ -0,0 +1,13 @@
+        background: var(--color-scale-white)
+        overflow-y: auto
+        border-radius: var(--borderRadius-medium)!important
+        margin: 5px
```
I'd prefer we place them somewhere like "ckeditor_primer_overrides" or `primer/_overrides` and pass a class to CKEditor so it becomes reusable. (Note to self: check with primer)

jjabari-op replied:
Good point! Done, it's now reusable!

oliverguenther commented on app/components/work_packages/activities_tab/journals/item_component/show.html.erb:
```diff
@@ -0,0 +1,15 @@
+  component_wrapper do
+    if journal.notes.present? || (journal.details.any? && !journal.initial?)
+      flex_layout do |journal_container|
+        if journal.notes.present?
```
I don't get this nesting of statements. If you have `journal.notes.present?` in the inner, you don't need the other conditions?

jjabari-op replied:
Done

oliverguenther commented on app/components/work_packages/activities_tab/journals/new_component.sass:
```diff
@@ -0,0 +1,28 @@
+        margin: 5px
+    // overwrites margin bottom defined in _ckeditor.sass:13
+    opce-ckeditor-augmented-textarea .op-ckeditor--wrapper
+        margin-bottom: 0px
```
Same as earlier, needs to be a BEM module and overrides should be generalized

jjabari-op replied:
Done

oliverguenther commented:
- [x] With a long name, the part after it is broken up in a weird way

<img width="586" alt="Bildschirmfoto 2024-08-27 um 16 08 11" src="https://github.com/user-attachments/assets/29792899-b446-4d94-8f0e-9289754ea785">

oliverguenther commented on app/components/work_packages/activities_tab/shared_helpers.rb:
```diff
@@ -0,0 +1,58 @@
+          component_collection do |collection|
+            collection.with_component(Primer::Beta::Truncate.new(classes: "hidden-for-mobile")) do |component|
+              component.with_item(max_width: 180) do
+                user.name
```
Why use this truncation in favor of using an ellipsis? we went with ellipsis in other implementations, so I want to double check here why you're doing it

jjabari-op replied:
Done

oliverguenther commented on app/controllers/work_packages/activities_tab_controller.rb:
```diff
@@ -0,0 +1,247 @@
+    render(
+      WorkPackages::ActivitiesTab::IndexComponent.new(
+        work_package: @work_package,
+        filter: params[:filter]&.to_sym || :all
```
This line is copied around multiple times in the controller. Put it in a before action instead?

oliverguenther commented:
- [x] As an anonymous user in a public project, or member without comment permission, I see the comment form and can click submit, nothing happens (no errors either)

<img width="603" alt="Bildschirmfoto 2024-08-27 um 16 10 09" src="https://github.com/user-attachments/assets/de835f20-2607-4969-a06b-a7b120d641fa">

- [x] I can also see "quote reply" without permission to comment
<img width="542" alt="Bildschirmfoto 2024-08-27 um 16 10 45" src="https://github.com/user-attachments/assets/52488acc-4adf-4579-93ab-8205432500c0">

jjabari-op committed:
auto-reset filter when adding a comment which would not be visible, minor refactoring

oliverguenther commented on app/controllers/work_packages/activities_tab_controller.rb:
```diff
@@ -0,0 +1,247 @@
+  end
+
+  def edit
+    # check if allowed to edit at all
```
Is this is a todo or why is this comment there?

jjabari-op replied:
Done

oliverguenther commented on app/controllers/work_packages/activities_tab_controller.rb:
```diff
@@ -0,0 +1,247 @@
+    if call.success? && call.result
+      update_item_component(call.result, state: :show, filter: params[:filter]&.to_sym || :all)
+    end
+    # TODO: handle errors
```
TODO :)

jjabari-op replied:
Done

oliverguenther commented on app/controllers/work_packages/activities_tab_controller.rb:
```diff
@@ -0,0 +1,247 @@
+  private
+
+  def find_work_package
+    @work_package = WorkPackage.find(params[:work_package_id])
```
All these `find...` methods need error handling for the turbo stream case. In a regular call this would look like:

```ruby
def find_work_package
  @work_package = ....
rescue ActiveRecord::RecordNotFound
  render_404
end
```

We need something similar here, maybe to return a toast?

jjabari-op replied:
Done

oliverguenther commented on app/controllers/work_packages/activities_tab_controller.rb:
```diff
@@ -0,0 +1,247 @@
+  end
+
+  def find_project
+    @project = @work_package.project
```
same as above: error handling

jjabari-op replied:
Done

oliverguenther commented on app/controllers/work_packages/activities_tab_controller.rb:
```diff
@@ -0,0 +1,247 @@
+  end
+
+  def find_journal
+    @journal = Journal.find(params[:id])
```
same as above: error handling

jjabari-op replied:
Done

oliverguenther commented on app/controllers/work_packages/activities_tab_controller.rb:
```diff
@@ -0,0 +1,247 @@
+      journals = journals.where.not(notes: "")
+    end
+
+    journals.where("updated_at > ?", last_update_timestamp).find_each do |journal|
```
Wouldn't it be easier to just rerender all at this point?

jjabari-op replied:
Then we would potentially loose an edit state

oliverguenther commented on app/controllers/work_packages/activities_tab_controller.rb:
```diff
@@ -0,0 +1,247 @@
+          filter:
+        )
+      )
+      # TODO: is it possible to loose an edit state this way?
```
Open TODO

oliverguenther commented:
- [x] Drag and Drop an into the comment field does no longer open it.

[Kapture 2024-08-27 at 16.16.15.webm](https://github.com/user-attachments/assets/72ff847b-521f-423e-87f1-5a4ca9173ae6)

oliverguenther commented on config/initializers/permissions.rb:
```diff
@@ -217,6 +217,8 @@
                        work_packages: %i[show index],
                        work_packages_api: [:get],
                        "work_packages/reports": %i[report report_details],
+                       "work_packages/activities_tab": %i[index create update_streams edit cancel_edit update
```
`create edit update`  etc shouldn't be part of the view_permission

jjabari-op replied:
Done

jjabari-op committed:
text adjustments as requested by Parimal

oliverguenther commented on frontend/src/app/features/work-packages/components/wp-edit/work-package-changeset.ts:
```diff
@@ -11,6 +11,13 @@ export class WorkPackageChangeset extends ResourceChangeset<WorkPackageResource>
+    // TODO: emit event when change is persisted
+    // currently the event might be fired too early as it only reflects the client side change
+    document.dispatchEvent(
+      new CustomEvent('work-package-updated'),
```
Instead of throwing around events, you could also get the controller and call actions on it: https://stimulus.hotwired.dev/reference/controllers#directly-invoking-other-controllers

jjabari-op replied:
I wasn't aware this is possible for communicating from angular to stimulus. Additionally, I don't see a big issue in using the event approach in this case. Do you much prefer not to use an event here?

oliverguenther commented on frontend/src/app/shared/components/editor/components/ckeditor/op-ckeditor.component.ts:
```diff
@@ -247,7 +247,7 @@ export class OpCkeditorComponent implements OnInit, OnDestroy {
           debugLog('Sending save request from CKEditor.');
           this.saveRequested.emit();
-          evt.stop();
+          // evt.stop();
```
??

jjabari-op replied:
Fixed

oliverguenther commented on frontend/src/global_styles/layout/work_packages/_full_view.sass:
```diff
@@ -79,7 +79,7 @@
       grid-template-rows: auto auto 1fr
       height: 100%
-      padding: 5px 0 10px 15px
+      padding: 5px 0 0px 15px
```
This will change padding on other tabs as well

jjabari-op replied:
Fixed

oliverguenther commented on frontend/src/stimulus/controllers/dynamic/work-packages/activities-tab/index.controller.ts:
```diff
@@ -0,0 +1,422 @@
+  private setLocalStorageKey() {
+    // scoped by user id in order to avoid data leakage when a user logs out and another user logs in on the same browser
+    // TODO: when a user logs out, the data should be removed anyways in order to avoid data leakage
+    this.localStorageKey = `work-package-${this.workPackageIdValue}-rescued-editor-data-${this.userIdValue}`;
```
we have a generalized approach for this now using the autosave feature of CKEditor. But it doesn't engage directly on beforeunload, but rather after a delay of inputting it. So it might still cause data loss if you're switching back and forth quickly. Whether it's worth to keep this in here too, I do not know

https://github.com/opf/openproject/pull/16476

jjabari-op replied:
Good to see a generalized approach for recovering client side ckeditor data. I would argue that in the case of the activity tab, it is quite likely that a user (accidentally) switches a tab and would still be confused why the unsaved data is gone. Recovering the data with the approach you're mentioning is then possible but IMO not convenient enough. As a user I would much prefer not to recover my data manually. This is especially important as a tab switch with the angular implementation of the activities did not lead to a ckeditor data loss. My intention was to keep this behavior. Thoughts?

oliverguenther commented on frontend/src/stimulus/controllers/dynamic/work-packages/activities-tab/index.controller.ts:
```diff
@@ -0,0 +1,422 @@
+            this.scrollJournalContainer(this.journalsContainerTarget, true);
+          }
+        }
+      }, 100);
```
This delay is rather arbitrary and causing a noticeable delay, also the scrolling duration is very long in my opinion. Please double check with product

jjabari-op replied:
The timeout is required in order to wait for `Turbo.renderStreamMessage(text)` to have processed and rendered the received stream updates on the actual DOM. `Turbo.renderStreamMessage(text)` does not return a promise or such as far as I investigated. One could use the connect callback of the added item(s) in order to identify when they have been added to the DOM and then trigger the scroll. But this would increase the code complexity quite a bit IMO. Thus I chose to use a timeout of 100ms which should be enough for almost all devices to apply the DOM changes. Additionally I did not see that a delay was noticeable.

oliverguenther commented on frontend/src/stimulus/controllers/dynamic/work-packages/activities-tab/index.controller.ts:
```diff
@@ -0,0 +1,422 @@
+    if (scrollableContainer) {
+      setTimeout(() => {
+        scrollableContainer.scrollTop = scrollableContainer.scrollHeight;
+      }, 400);
```
Again, this timeout is rather long and arbitrary. Why 400ms?

jjabari-op replied:
This was taken from the original scroll implementation of the activity tab, but it's removed completely now

oliverguenther commented on frontend/src/stimulus/controllers/dynamic/work-packages/activities-tab/index.controller.ts:
```diff
@@ -0,0 +1,422 @@
+    return this.element.querySelector('#input-container');
+  }
+
+  // TODO: get rid of static width value and reach for a more CSS based solution
+  private isMobile():boolean {
+    return window.innerWidth < 1279;
+  }
+
+  // TODO: get rid of static width value and reach for a more CSS based solution
+  private isSmViewPort():boolean {
+    return window.innerWidth < 543;
+  }
+
+  // TODO: get rid of static width value and reach for a more CSS based solution
+  private isMdViewPort():boolean {
+    return window.innerWidth >= 543 && window.innerWidth < 1279;
+  }
```
Agreed, do not use fixed size values here and instead use our CSS breakpoints

jjabari-op replied:
I hope, we can get rid of all of that when we have the standalone activity tab on mobile for the workpackage page as well. We then have the same constraints on all viewports, which would be great!

oliverguenther commented on frontend/src/stimulus/controllers/dynamic/work-packages/activities-tab/index.controller.ts:
```diff
@@ -0,0 +1,422 @@
+        // taken from op-ck-editor.component.ts
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-enum-comparison
+        if ((data.ctrlKey || data.metaKey) && data.keyCode === KeyCodes.ENTER) {
+          void this.onSubmit();
```
Why did you need to duplicate this code here? can't we use an output from the op-ckeditor component?

jjabari-op replied:
Done

oliverguenther commented on frontend/src/stimulus/controllers/dynamic/work-packages/activities-tab/index.controller.ts:
```diff
@@ -0,0 +1,422 @@
+        this.adjustJournalContainerMargin();
+        event.stop();
+      },
+      { priority: 'highest' },
```
Same as above, duplicating op-ckeditor code

jjabari-op replied:
Done

oliverguenther commented on frontend/src/stimulus/controllers/dynamic/work-packages/activities-tab/index.controller.ts:
```diff
@@ -0,0 +1,422 @@
+        }, 0);
+      },
+      { priority: 'highest' },
+    );
```
Same as above, duplicating op-ckeditor code. I'd like to see this remain in the op-ckeditor custom element so it can be reused.

jjabari-op replied:
Done

oliverguenther commented on frontend/src/stimulus/controllers/dynamic/work-packages/activities-tab/index.controller.ts:
```diff
@@ -0,0 +1,422 @@
+      },
+      credentials: 'same-origin',
+    });
+  }
```
We should extract this into the  TurboRequests service (src/app/core/turbo/turbo-requests.service.ts) , except for the token, it's already there. It's still valid to use Angular services for this use case.

jjabari-op replied:
Done

oliverguenther commented on frontend/src/stimulus/controllers/dynamic/work-packages/activities-tab/index.controller.ts:
```diff
@@ -0,0 +1,422 @@
+
+    const response = await this.fetchWithCSRF(this.formTarget.action, 'POST', formData);
+
+    if (response.ok) {
+      this.setLastUpdateTimestamp();
+      const text = await response.text();
+      Turbo.renderStreamMessage(text);
+
+      if (this.journalsContainerTarget) {
+        this.clearEditor();
+        if (this.isMobile()) {
+          this.hideEditorIfEmpty();
+        } else {
+          this.focusEditor();
+        }
+        if (this.journalsContainerTarget) {
+          this.journalsContainerTarget.style.marginBottom = '';
+          this.journalsContainerTarget.classList.add('with-input-compensation');
+        }
+        setTimeout(() => {
+          this.scrollJournalContainer(
+            this.journalsContainerTarget,
+            this.sortingValue === 'asc',
+          );
+          if (this.isMobile()) {
+            this.scrollInputContainerIntoView(300);
+          }
+        }, 10);
+      }
+    }
+  }
```
This block is very hard to read, could you extract it into a method to remove the nesting of statements?

jjabari-op replied:
Done

oliverguenther commented on spec/features/activities/work_package/activities_spec.rb:
```diff
@@ -0,0 +1,758 @@
+
+require "spec_helper"
+
+RSpec.describe "Work package activity", :js, :with_cuprite, with_flag: { primerized_work_package_activities: true } do
```
This spec is missing all other permission checks (i.e, having only view, none permission etc.)

jjabari-op replied:
Done

oliverguenther reviewed (changes requested):
@jjabari-op A very quick, rough, preliminary review as I feel like this review should have been made by someone of the same stream. There's still lots of things marked as TODO that we'd need to address, most and foremost:

- Permissions are not properly checked for all actions, resulting in more things being shown that are or should not be possible
- Too specific CSS, non-BEM CSS and no generalization for some code that should be there
- Duplicated code at certain places, especially taking away from the op-ckeditor component that should be in there instead.

jjabari-op committed:
fixed margin bottom for desc sorting

jjabari-op committed:
fixed permissions, handle permissions on UI, added permission specs

jjabari-op committed:
added and refactore permission specs

jjabari-op committed:
WIP: adding controller spec and exception handling

jjabari-op committed:
added exception handling and controller specs

lindenthal commented:
hi @jjabari-op,

I just had a brief look on iOS. There were a couple of problems. Can you please also test it small screens?

Thanks
Niels

jjabari-op commented:
> hi @jjabari-op,
> 
> I just had a brief look on iOS. There were a couple of problems. Can you please also test it small screens?
> 
> Thanks Niels

Hi @lindenthal! Sorry for that, some optimizations I've implemented recently broke some mobile behavior without me noticing it. Will fix it asap!

Best, Jonas

jjabari-op committed:
refactored CSS and applied BEM notations

jjabari-op committed:
fixed specs after refactoring

jjabari-op committed:
fixed stem for filter states

jjabari-op committed:
mobile UX fixes

jjabari-op committed:
Merge branch 'dev' into feature/54733-primerise-the-activity-panel

jjabari-op committed:
temporarly removed meta data as its consuming too much width with long date format settings, clarification with product team needed

jjabari-op committed:
ckeditor fixes due to more toolbar items, open editor when dragover, ckeditor still not accepting dropping files

jjabari-op committed:
fixed positioning issues caused by anchor scrolls via url and click on journal version

jjabari-op committed:
added support for retracted journal entries

jjabari-op committed:
disabling some rubocop checks as the permission specs seem not to follow them in general

jjabari-op committed:
WIP: implementing review feedback from @oliverguenther

jjabari-op committed:
make ckeditor primer adjustments reusable following input from @oliverguenther

jjabari-op committed:
WIP implementing review feedback from Oliver

jjabari-op committed:
added generic error handling for initial and subsequent requests

jjabari-op committed:
moved ckeditor code as requested by @oliverguenther

jjabari-op committed:
js refactoring as requested by @oliverguenther

jjabari-op committed:
extended permission controller specs

jjabari-op committed:
rubocop fixes

jjabari-op committed:
styling fixes

jjabari-op committed:
fixed specs, eslint errors and prevent multi form submission

jjabari-op committed:
Merge branch 'dev' into feature/54733-primerise-the-activity-panel

jjabari-op committed:
fixed flaky specs

jjabari-op committed:
fixed flaky specs and styling fixes

jjabari-op committed:
refactored http request as requested by @oliverguenther

jjabari-op committed:
note future required code maintenance on js based view port checks

jjabari-op committed:
fix stem rendering as requested by @psatyal

jjabari-op committed:
adjusted notification dot color as requested by @psatyal

jjabari-op committed:
fixed scrolling behaviour on mobile screens

jjabari-op committed:
restrict polling interval env to test env

jjabari-op committed:
fixed ellipsis

jjabari-op committed:
trying to avoid flaky spec

jjabari-op committed:
Merge branch 'dev' into feature/54733-primerise-the-activity-panel

jjabari-op committed:
trying to fix flaky spec through a more dynamic waiting approach

jjabari-op committed:
removed obsolete sleeps

jjabari-op committed:
Merge branch 'dev' into feature/54733-primerise-the-activity-panel

jjabari-op committed:
fixing auto-scrollings due to content wrapper changes resolving anchor menu positions on mobile

jjabari-op committed:
only perform scrolling when really required

jjabari-op committed:
fixed eslint error

dominic-braeunlein commented:
**Auto generated description (might contain errors):**

**Context**
This pull request addresses the need to "primerize" the activity panel in the OpenProject work package view. The goal is to enhance the user experience by adopting Primer design components, improving the UI consistency, and making the activity panel more responsive and visually appealing. The changes are being made to modernize the UI and align with OpenProject's ongoing UI/UX improvements, as well as to introduce better error handling and user interaction improvements within the activity panel.

**Description**
The task involves updating the activity panel to use Primer components, which is part of OpenProject’s broader move towards a more modern, consistent, and accessible UI. The activity panel is a crucial part of the work package view, where users can see the history of changes, comments, and other activities related to a work package. The changes include refactoring existing components to use Primer's design system, adding new components for error handling, and improving the interaction flow for adding comments and filtering activities.

Steps to accomplish this task include:
1. Refactoring the existing activity panel to use Primer components for layout and interaction.
2. Adding new components for handling errors and empty states.
3. Implementing new components for filtering and sorting activities.
4. Ensuring the new components are responsive and accessible across different screen sizes.
5. Integrating the changes into the existing work package view and ensuring compatibility with other parts of the system.

**Changes in the codebase**
1. **New Components**:
   - **Error Handling**: `ErrorFrameComponent` and `ErrorStreamComponent` handle error messages gracefully using Primer's banner component.
   - **Activity Panel**: `IndexComponent`, `ItemComponent`, `DetailsComponent`, `NewComponent`, and `FilterAndSortingComponent` manage the display of activities, comments, and filtering/sorting options. These components use Primer’s layout system, including flex layouts, buttons, and action menus.
   - **Empty State**: `EmptyComponent` displays a friendly message when there are no activities to show.
   - **Form Handling**: `NewComponent` manages the form for adding new comments, with a toggle between the button and form views.
   - **Shared Helpers**: A helper module (`SharedHelpers`) was introduced for common tasks like rendering user names and links.

2. **Styling**:
   - New SCSS files were added for each component to handle the layout and styling, ensuring that the components are responsive and follow the Primer design guidelines.
   - Specific adjustments were made for mobile and desktop views to ensure the UI works well across different screen sizes.

3. **Backend Logic**:
   - The backend logic for handling activity filtering and sorting was updated. The `FilterAndSortingComponent` allows users to filter activities by "all," "only changes," or "only comments" and sort them in ascending or descending order.

4. **Integration with Existing System**:
   - The new activity panel components are integrated into the work package view, replacing the old activity panel. The integration ensures that the new components work seamlessly with the existing functionality, such as permissions for adding comments and notifications.

**Changes outside the codebase**
1. **Environment Variables**: 
   - A new environment variable `OPENPROJECT_FEATURE_PRIMERIZED_WORK_PACKAGE_ACTIVITIES_ACTIVE` was added to enable or disable the primerized activity panel in different environments.
   
2. **Third-Party Services**: 
   - No changes were made to third-party services.

3. **Infrastructure**: 
   - No changes were made to the infrastructure configuration.

**Additional Information**
- The polling interval for updating the activity stream is set to 10 seconds by default, which is configurable in the test environment.
- The UI has been designed to be accessible and responsive, with special attention to mobile layouts.
- The use of Turbo Streams ensures that the activity panel can update dynamically without requiring a full page refresh.

**Work Package**
For more details, you can refer to the work package [#54733](https://community.openproject.org/projects/openproject/work_packages/54733), which outlines the need for primerizing the activity panel and provides additional context on the expected outcomes and design goals.

**Code Challenge**
1. **Error Handling**: The `ErrorFrameComponent` and `ErrorStreamComponent` handle errors well by displaying a danger banner when an error occurs. However, it is important to ensure that the error messages are properly sanitized to avoid potential XSS vulnerabilities, especially since user-generated content could be involved.
   
2. **Polling Interval**: The polling interval is hardcoded to 10 seconds, which might be too frequent for some use cases. It would be beneficial to allow this to be configurable through the environment or user preferences in production environments as well.

3. **Journal Sorting Logic**: The sorting logic for journals is based on the user's preference, defaulting to "desc." However, if a user does not have a preference set, it might be useful to provide a more explicit fallback mechanism or prompt the user to set their preference.

4. **Form Toggle**: The toggle between the comment button and the form works well, but there might be edge cases where the form does not render correctly if the DOM is manipulated too quickly (e.g., during rapid clicking). It might be useful to debounce the click event to prevent such issues.

5. **Empty State**: The `EmptyComponent` provides a good user experience when no activities are present. However, it might be useful to implement a more detailed message or action (e.g., a prompt to add the first comment) to encourage user interaction.

jjabari-op committed:
update workpackage attributes on main workpackage section when changeset was polled

jjabari-op committed:
fixed attachement links in changesets

jjabari-op committed:
added missing labels and aria-labels, adjusted username width in order not break UI on smallest possible side pane width

jjabari-op committed:
fixed stem styling

jjabari-op committed:
rename sorting labels as requested

jjabari-op committed:
removed latest activties from overview tab as requested by @wielinde in WP #58017

jjabari-op committed:
update the count in the primerized activty tab header when new activity is polled

jjabari-op committed:
Merge branch 'dev' into feature/54733-primerise-the-activity-panel

jjabari-op committed:
prototypical implementation of the toggle singel notifications feature, specs and refactoring required

jjabari-op committed:
prototypical changes to the notification center and its behaviour when new notifications are polled

jjabari-op committed:
decrease polling interval via env var for pullpreview

lindenthal commented:
Hi @jonasjabari  and @wielinde,

I just added a comment on mobile. Currently it requires two clicks/taps to enter text. Here in GitHub only one click is required. Can you please check?
Thanks
Niels

lindenthal commented:
![image](https://github.com/user-attachments/assets/3be6b978-bdd7-420c-8132-2416f8fa4c0c)

Hi @jonasjabari and @wielinde 
I just found another problem with embedded work package macros on mobile.
Can you please check?
Thanks
Niels

jjabari-op commented:
Thank you for your feedback @lindenthal 
I have both things on my list now!

jjabari-op committed:
Revert "prototypical implementation of the toggle singel notifications feature, specs and refactoring required"

This reverts commit 66aa22978ba5a57c4ddc5fc5455e556f8b6b0ab7.

jjabari-op committed:
reverted browser notification experiment

jjabari-op committed:
not using a flash message anymore as decided with product team, workpackage relation check not required anymore

jjabari-op committed:
adjusted notification bubble color in favicon as requested

jjabari-op committed:
cleanup

jjabari-op committed:
enable immediate notification center and bell update through ianCenterService usage and enhancement

jjabari-op committed:
display labels as requested

jjabari-op committed:
Merge branch 'dev' into feature/54733-primerise-the-activity-panel

jjabari-op committed:
put changes in notification center behind feature flag as well

jjabari-op committed:
fixed too short ellipsis

jjabari-op committed:
trying to fix flaky spec

jjabari-op committed:
fix eslint issues on CI

jjabari-op committed:
fix rubocop issues on CI

jjabari-op committed:
fix scroll behavior on mobile

jjabari-op committed:
reapplied refactoring previously reverted

akabiru committed:
Merge branch 'dev' into feature/54733-primerise-the-activity-panel

akabiru committed:
chore[Op#54733]: add missing newlines

akabiru committed:
fix[Op#54733]: restore old concept to bind to `notificationCountIncreased`

In the original flow, which still exists- a blue banner is shown anytime there are new notifications
based on unreadCountIncreased$ - however, when based on a generic unreadCountChanged$ it is not
dismissible

akabiru committed:
chore[Op#54733]: extract feature flag check to function

akabiru committed:
chore[Op#54733]: remove red dot favicon- out of scope

akabiru committed:
chore[Op#54733]: remove leftover var

akabiru committed:
chore[Op#54733]: resolve eslint missing comma

Kharonus reviewed (approved):
LGTM in the context of the pair review with @akabiru 

- the PR is way too big to get a good overview of all introduced decisions and code
- there are some weird architectural patterns (mixin angular, stimulus and rails) I'd challenge, if there is time
- We agreed, that the right time of raising the challenges are, when the angular part is partially deconstructed from this part of the product -> then we need to take the time and rethink some of the decision taken here
